File: testdata/languages/php/sample.php
<?php
declare(strict_types=1);

namespace App;

enum Status: string { case Pending = 'pending'; case Running = 'running'; case Done = 'done'; }

interface Repository {
    public function find(string $id): ?Entity;
    public function save(Entity $e): void;
}

final class Entity {
    public function __construct(
        public readonly string $id,
        public string $name,
        public Status $status = Status::Pending,
        public array $meta = []
    ) {}
    public function isDone(): bool { return $this->status === Status::Done; }
    public function toArray(): array { return ['id'=>$this->id,'name'=>$this->name,'status'=>$this->status->value]; }
}

final class Service {
    /** @var array<string,Entity> */
    private array $cache = [];
    public function __construct(private Repository $repo) {}
    public function execute(string $id): Entity {
        if (isset($this->cache[$id])) return $this->cache[$id];
        $e = $this->repo->find($id);
        if ($e === null) throw new \RuntimeException("not found $id");
        $e->status = Status::Running;
        $this->repo->save($e);
        $this->cache[$id] = $e;
        return $e;
    }
    public function validate(string $name): bool { return $name !== ''; }
}

final class Controller {
    public function __construct(private Service $svc) {}
    public function index(string $id): string {
        $e = $this->svc->execute($id);
        return json_encode($e->toArray());
    }
}

File: testdata/languages/rust/sample.rs
use std::collections::HashMap;
use std::sync::{Arc, Mutex};

#[derive(Debug, Clone, PartialEq)]
pub enum Status { Pending, Running, Done }

#[derive(Debug, Clone)]
pub struct Entity {
    pub id: String,
    pub name: String,
    pub status: Status,
    pub meta: HashMap<String, String>,
}

impl Entity {
    pub fn new(id: impl Into<String>, name: impl Into<String>) -> Self {
        Self { id: id.into(), name: name.into(), status: Status::Pending, meta: HashMap::new() }
    }
    pub fn is_done(&self) -> bool { self.status == Status::Done }
}

pub trait Repository: Send + Sync {
    fn find(&self, id: &str) -> Option<Entity>;
    fn save(&self, e: Entity) -> Result<(), String>;
}

pub struct Service {
    repo: Arc<dyn Repository>,
    cache: Mutex<HashMap<String, Entity>>,
}

impl Service {
    pub fn new(repo: Arc<dyn Repository>) -> Self {
        Self { repo, cache: Mutex::new(HashMap::new()) }
    }
    pub fn execute(&self, id: &str) -> Result<Entity, String> {
        if let Some(cached) = self.cache.lock().unwrap().get(id).cloned() {
            return Ok(cached);
        }
        let mut e = self.repo.find(id).ok_or_else(|| format!("not found {id}"))?;
        e.status = Status::Running;
        self.repo.save(e.clone())?;
        self.cache.lock().unwrap().insert(id.to_string(), e.clone());
        Ok(e)
    }
}

pub struct Worker {
    pub id: u64,
}

impl Worker {
    pub fn new(id: u64) -> Self { Worker { id } }
    pub fn run(&self) -> Result<(), String> { Ok(()) }
}

#[cfg(test)]
mod tests {
    use super::*;
    #[test] fn test_entity() { let e = Entity::new("1", "demo"); assert!(!e.is_done()); }
}
